SpatialGDK/Source/SpatialGDK/Private/Utils/ComponentReader.cpp

```diff
@@ -187,184 +179,224 @@ void ComponentReader::ApplySchemaObject(Schema_Object* ComponentObject, UObject&
+		bOutReferencesChanged = ApplySchemaObjectData.bOutReferencesChanged;
+	}
+
+	Channel.RemoveRepNotifiesWithUnresolvedObjs(RepNotifies, *Replicator->RepLayout, RootObjectReferencesMap, &Object);
+
+	Channel.PostReceiveSpatialUpdate(&Object, RepNotifies, PropertySpanIds);
+}
+
+void ComponentReader::ApplySchemaObjectFields(ApplySchemaObjectDataStruct& ApplySchemaObjectData)
+{
+	FObjectReplicator* Replicator = ApplySchemaObjectData.Replicator;
+	UObject& Object = ApplySchemaObjectData.Object;
+	USpatialActorChannel& Channel = ApplySchemaObjectData.Channel;
+	Worker_ComponentId ComponentId = ApplySchemaObjectData.ComponentId;
+
+	TUniquePtr<FRepState>& RepState = Replicator->RepState;
+	TArray<FRepLayoutCmd>& Cmds = Replicator->RepLayout->Cmds;
+	TArray<FHandleToCmdIndex>& BaseHandleToCmdIndex = Replicator->RepLayout->BaseHandleToCmdIndex;
+	TArray<FRepParentCmd>& Parents = Replicator->RepLayout->Parents;
+
+	const Worker_EntityId EntityId = Channel.GetEntityId();
+	const bool bEventTracerEnabled = EventTracer != nullptr;
+
+	const bool bIsAuthServer = Channel.IsAuthoritativeServer();
+	const bool bAutonomousProxy = Channel.IsClientAutonomousProxy();
+	const bool bIsClient = NetDriver->GetNetMode() == NM_Client;
+	const FSpatialConditionMapFilter ConditionMap(&Channel, bIsClient);
+
+	for (uint32 FieldId : ApplySchemaObjectData.UpdatedIds)
+	{
+		// FieldId is the same as rep handle
+		if (FieldId == 0 || (int)FieldId - 1 >= BaseHandleToCmdIndex.Num())
 		{
-			// FieldId is the same as rep handle
-			if (FieldId == 0 || (int)FieldId - 1 >= BaseHandleToCmdIndex.Num())
-			{
-				UE_LOG(LogSpatialComponentReader, Error,
-					   TEXT("ApplySchemaObject: Encountered an invalid field Id while applying schema. Object: %s, Field: %d, Entity: "
-							"%lld, Component: %d"),
-					   *Object.GetPathName(), FieldId, Channel.GetEntityId(), ComponentId);
-				continue;
-			}
+			UE_LOG(LogSpatialComponentReader, Error,
+				   TEXT("ApplySchemaObject: Encountered an invalid field Id while applying schema. Object: %s, Field: %d, Entity: "
+						"%lld, Component: %d"),
+				   *Object.GetPathName(), FieldId, Channel.GetEntityId(), ComponentId);
+			continue;
+		}
 
-			int32 CmdIndex = BaseHandleToCmdIndex[FieldId - 1].CmdIndex;
-			const FRepLayoutCmd& Cmd = Cmds[CmdIndex];
-			const FRepParentCmd& Parent = Parents[Cmd.ParentIndex];
-			int32 ShadowOffset = Cmd.ShadowOffset;
+		int32 CmdIndex = BaseHandleToCmdIndex[FieldId - 1].CmdIndex;
+		const FRepLayoutCmd& Cmd = Cmds[CmdIndex];
+		const FRepParentCmd& Parent = Parents[Cmd.ParentIndex];
 
-			if (NetDriver->IsServer() || ConditionMap.IsRelevant(Parent.Condition))
-			{
-				// This is mostly copied from ReceivePropertyHelper in RepLayout.cpp
-				auto GetSwappedCmd = [&Cmd, &Cmds, &Parents, bIsAuthServer, &Replicator, &Channel, &Parent]() -> const FRepLayoutCmd& {
+		bool IsSpecialCaseProperty = SpecialCaseProperties.Contains(Cmd.Property->GetFName());
```

Comment:
Mostly, this big change is just to break out part of ApplySchemaObject into a separate function, ApplySchemaFields -- the main bit is right here, checking if a property is one of the special cases so that we can process it during the first call of ApplySchemaFields, and then call ApplySchemaFields again on the rest of the properties.